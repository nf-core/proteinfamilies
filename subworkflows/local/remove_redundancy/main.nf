/*
    REMOVAL OF REDUNDANT SEQUENCES AND FAMILIES
*/

include { EXTRACT_FAMILY_REPS       } from '../../../modules/local/extract_family_reps/main'
include { CAT_CAT                   } from '../../../modules/nf-core/cat/cat'
include { HMMER_HMMSEARCH           } from '../../../modules/nf-core/hmmer/hmmsearch/main'
include { REMOVE_REDUNDANT_FAMS     } from '../../../modules/local/remove_redundant_fams/main'
include { FILTER_NON_REDUNDANT_HMMS } from '../../../modules/local/filter_non_redundant_hmms/main'
include { EXECUTE_CLUSTERING        } from '../../../subworkflows/local/execute_clustering'
include { REMOVE_REDUNDANT_SEQS     } from '../../../modules/local/remove_redundant_seqs/main'
include { ALIGN_SEQUENCES           } from '../../../subworkflows/local/align_sequences'

workflow REMOVE_REDUNDANCY {
    take:
    msa   // tuple val(meta), path(fas)
    fasta // tuple val(meta), path(fasta)
    hmm   // tuple val(meta), path(hmm)

    main:
    ch_versions = Channel.empty()

    if (params.remove_family_redundancy) {
        ch_msa = msa
            .map { meta, aln -> [[id: meta.id], aln] }
            .groupTuple(by: 0)
        EXTRACT_FAMILY_REPS( ch_msa )
        ch_versions = ch_versions.mix( EXTRACT_FAMILY_REPS.out.versions )

        ch_hmm = hmm
            .map { meta, model -> [[id: meta.id], model] }
            .groupTuple(by: 0)
        CAT_CAT( ch_hmm )
        ch_versions = ch_versions.mix( CAT_CAT.out.versions )

        ch_input_for_hmmsearch = CAT_CAT.out.file_out
            .combine(EXTRACT_FAMILY_REPS.out.fasta, by: 0)
            .map { meta, model, seqs -> [meta, model, seqs, false, false, true] }

        HMMER_HMMSEARCH( ch_input_for_hmmsearch )
        ch_versions = ch_versions.mix( HMMER_HMMSEARCH.out.versions )

        fasta = fasta
            .map { meta, fas -> [[id: meta.id], fas] }
            .groupTuple(by: 0)

        // Join to ensure in sync
        ch_input_for_fam_removal = EXTRACT_FAMILY_REPS.out.map
            .join(HMMER_HMMSEARCH.out.domain_summary)
            .join(fasta)
            .multiMap { meta, map, domtbl, seqs ->
                map: [meta, map]
                domtbl: [meta, domtbl]
                seqs: [meta, seqs]
            }

        REMOVE_REDUNDANT_FAMS( ch_input_for_fam_removal.map, ch_input_for_fam_removal.domtbl, ch_input_for_fam_removal.seqs, params.hmmsearch_family_length_threshold )
        ch_versions = ch_versions.mix( REMOVE_REDUNDANT_FAMS.out.versions )
        fasta = REMOVE_REDUNDANT_FAMS.out.fasta

        // Join to ensure in sync
        ch_input_for_hmm_filtering = fasta
            .join(ch_hmm)
            .multiMap { meta, seqs, models ->
                seqs: [meta, seqs]
                models: [meta, models]
            }
        FILTER_NON_REDUNDANT_HMMS( ch_input_for_hmm_filtering.seqs, ch_input_for_hmm_filtering.models )
        ch_versions = ch_versions.mix( FILTER_NON_REDUNDANT_HMMS.out.versions )

        fasta = fasta
            .transpose()
            .map { meta, file ->
                [[id: meta.id, chunk: file.getSimpleName().split('_')[-1]], file]
            }
    }

    if (params.remove_sequence_redundancy) {
        EXECUTE_CLUSTERING( fasta, params.clustering_tool )
        ch_versions = ch_versions.mix( EXECUTE_CLUSTERING.out.versions )

        REMOVE_REDUNDANT_SEQS( EXECUTE_CLUSTERING.out.clusters, EXECUTE_CLUSTERING.out.seqs )
        ch_versions = ch_versions.mix( REMOVE_REDUNDANT_SEQS.out.versions )

        ALIGN_SEQUENCES( REMOVE_REDUNDANT_SEQS.out.fasta, params.alignment_tool )
        ch_versions = ch_versions.mix( ALIGN_SEQUENCES.out.versions )
        msa = ALIGN_SEQUENCES.out.alignments
    }

    emit:
    versions = ch_versions
    msa      = msa
}
