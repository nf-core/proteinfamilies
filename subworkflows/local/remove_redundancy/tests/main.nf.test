nextflow_workflow {

    name "test Subworkflow REMOVE_REDUNDANCY"
    script "../main.nf"
    workflow "REMOVE_REDUNDANCY"

    test("fasta") {

        when {
            workflow {
                """
                input[0] = Channel.of(
                    [
                        [ id: 'chunk', chunk:1 ],
                        file(params.pipelines_testdata_base_path + 'test_data/subworkflows/chunk_1.clipkit', checkIfExists: true)
                    ],
                    [
                        [ id: 'chunk', chunk:2 ],
                        file(params.pipelines_testdata_base_path + 'test_data/subworkflows/chunk_2.clipkit', checkIfExists: true)
                    ],
                    [
                        [ id: 'chunk', chunk:3 ],
                        file(params.pipelines_testdata_base_path + 'test_data/subworkflows/chunk_3.clipkit', checkIfExists: true)
                    ]
                )
                input[1] = Channel.of(
                    [
                        [ id: 'chunk', chunk:1 ],
                        file(params.pipelines_testdata_base_path + 'test_data/subworkflows/chunk_1.sto.gz', checkIfExists: true)
                    ],
                    [
                        [ id: 'chunk', chunk:2 ],
                        file(params.pipelines_testdata_base_path + 'test_data/subworkflows/chunk_2.sto.gz', checkIfExists: true)
                    ],
                    [
                        [ id: 'chunk', chunk:3 ],
                        file(params.pipelines_testdata_base_path + 'test_data/subworkflows/chunk_3.sto.gz', checkIfExists: true)
                    ]
                )
                input[2] = Channel.of(
                    [
                        [ id: 'chunk', chunk:1 ],
                        file(params.pipelines_testdata_base_path + 'test_data/subworkflows/chunk_1.fasta.gz', checkIfExists: true)
                    ],
                    [
                        [ id: 'chunk', chunk:2 ],
                        file(params.pipelines_testdata_base_path + 'test_data/subworkflows/chunk_2.fasta.gz', checkIfExists: true)
                    ],
                    [
                        [ id: 'chunk', chunk:3 ],
                        file(params.pipelines_testdata_base_path + 'test_data/subworkflows/chunk_3.fasta.gz', checkIfExists: true)
                    ]
                )
                input[3] = Channel.of(
                    [
                        [ id: 'chunk', chunk:1 ],
                        file(params.pipelines_testdata_base_path + 'test_data/subworkflows/chunk_1.hmm.gz', checkIfExists: true)
                    ],
                    [
                        [ id: 'chunk', chunk:2 ],
                        file(params.pipelines_testdata_base_path + 'test_data/subworkflows/chunk_2.hmm.gz', checkIfExists: true)
                    ],
                    [
                        [ id: 'chunk', chunk:3 ],
                        file(params.pipelines_testdata_base_path + 'test_data/subworkflows/chunk_3.hmm.gz', checkIfExists: true)
                    ]
                )
                input[4] = true       // remove_family_redundancy
                input[5] = 0.9        // hmmsearch_family_length_threshold
                input[6] = true       // remove_sequence_redundancy
                input[7] = "linclust" // clustering_tool
                input[8] = "famsa"    // alignment_tool
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success},
                { assert snapshot(
                    workflow.out.fasta,
                    workflow.out.versions.collect { path(it).yaml }.unique()
                    ).match()}
            )
        }
    }

}
