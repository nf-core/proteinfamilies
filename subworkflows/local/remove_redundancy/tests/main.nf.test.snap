{
    "fasta": {
        "content": [
            [
                [
                    {
                        "id": "chunk",
                        "chunk": "1"
                    },
                    "chunk_1.faa:md5,6a54d12b222346b1df685867c0bdd151"
                ],
                [
                    {
                        "id": "chunk",
                        "chunk": "2"
                    },
                    "chunk_2.faa:md5,ebcda473387ebd3d1cdeb7179d4915d2"
                ]
            ],
            [
                {
                    "REMOVE_REDUNDANCY:IDENTIFY_REDUNDANT_FAMS": {
                        "python": "3.13.1",
                        "pandas": "2.3.0"
                    }
                },
                {
                    "REMOVE_REDUNDANCY:MMSEQS_FASTA_CLUSTER:MMSEQS_CREATETSV": {
                        "mmseqs": "17.b804f"
                    }
                },
                {
                    "REMOVE_REDUNDANCY:MMSEQS_FASTA_CLUSTER:MMSEQS_LINCLUST": {
                        "mmseqs": "17.b804f"
                    }
                },
                {
                    "REMOVE_REDUNDANCY:EXTRACT_FAMILY_REPS": {
                        "python": "3.13.1"
                    }
                },
                {
                    "REMOVE_REDUNDANCY:FILTER_NON_REDUNDANT_FASTA": {
                        "python": "3.13.1"
                    }
                },
                {
                    "REMOVE_REDUNDANCY:REMOVE_REDUNDANT_SEQS": {
                        "python": "3.13.1",
                        "biopython": 1.85
                    }
                },
                {
                    "REMOVE_REDUNDANCY:MMSEQS_FASTA_CLUSTER:MMSEQS_CREATEDB": {
                        "mmseqs": "17.b804f"
                    }
                },
                {
                    "REMOVE_REDUNDANCY:HMMER_HMMSEARCH": {
                        "hmmer": 3.4
                    }
                },
                {
                    "REMOVE_REDUNDANCY:FILTER_NON_REDUNDANT_SEED_MSA": {
                        "python": "3.13.1"
                    }
                },
                {
                    "REMOVE_REDUNDANCY:FILTER_NON_REDUNDANT_HMM": {
                        "python": "3.13.1"
                    }
                },
                {
                    "REMOVE_REDUNDANCY:ALIGN_SEQUENCES:FAMSA_ALIGN": {
                        "famsa": "2.4.1-45c9b2b (2025-05-09)"
                    }
                },
                {
                    "REMOVE_REDUNDANCY:FILTER_NON_REDUNDANT_FULL_MSA": {
                        "python": "3.13.1"
                    }
                },
                {
                    "REMOVE_REDUNDANCY:FIND_CONCATENATE": {
                        "find": "4.6.0",
                        "pigz": 2.8,
                        "coreutils": 9.4
                    }
                }
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.7"
        },
        "timestamp": "2025-09-26T10:22:32.046614612"
    }
}